# Project 3 Using Pandsas with used cars from eBay Kleinanzeigen dataset

The dataset was originally scraped and uploaded to Kaggle by user orgesleka. The original dataset isn't available on Kaggle anymore, but you can find it [here](https://data.world/data-society/used-cars-data).

Few modifications were already done from the original dataset:

- We sampled 50,000 data points from the full dataset, to ensure your code runs quickly in our hosted environment
- We dirtied the dataset a bit to more closely resemble what you would expect from a scraped dataset (the version uploaded to Kaggle was cleaned to be easier to work with)

The data dictionary provided with data is as follows:

- dateCrawled - When this ad was first crawled. All field-values are taken from this date.
- name - Name of the car.
- seller - Whether the seller is private or a dealer.
- offerType - The type of listing
- price - The price on the ad to sell the car.
- abtest - Whether the listing is included in an A/B test.
- vehicleType - The vehicle Type.
- yearOfRegistration - The year in which the car was first registered.
- gearbox - The transmission type.
- powerPS - The power of the car in PS.
- model - The car model name.
- kilometer - How many kilometers the car has driven.
- monthOfRegistration - The month in which the car was first registered.
- fuelType - What type of fuel the car uses.
- brand - The brand of the car.
- notRepairedDamage - If the car has a damage which is not yet repaired.
- dateCreated - The date on which the eBay listing was created.
- nrOfPictures - The number of pictures in the ad.
- postalCode - The postal code for the location of the vehicle.
- lastSeenOnline - When the crawler saw this ad last online.

**The aim of this project** is to clean the data and analyze the included used car listings. As well as become familiar with some of the unique benefits jupyter notebook provides for pandas.

# Part 1 - Importing Data

In [1]:
import pandas as pd
import numpy as np

# READING A CSV IN WITH A SPECIFIC ENCODING (most common 'Latin-1', 'UTF-8', 'Windows-1251')
autos = pd.read_csv('autos.csv', encoding='Latin-1')

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [3]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

Our dataset contains:
- 20 columns, most of which are stored as strings. 
- There are 5 columns with null values, but no columns have more than ~20% null values. 
- There are 3 columns that contain dates stored as strings.

We'll start by cleaning the column names to make the data easier to work with.

# Part 2 - Cleaning Data

## 2.1 Cleaning headers
The column names use camelcase instead of Python's preferred snakecase, which means we can't just replace spaces with underscores. 

Let's do the following:
1. reword some of the column names based on the data dictionary to be more descriptive.
2. convert the column names from [camelcase](https://en.wikipedia.org/wiki/Camel_case) to [snakecase](https://en.wikipedia.org/wiki/Snake_case) 

We can use following:
- ```DataFrame.rename(columns={'MANUfacturer' : 'manufacturer'}, inplace=True)``` enaming An Existing Column
- ```DataFrame.columns``` to get list of columns, as well as modify them
- ```Series.str.lower()``` to convert camelcase to snakecase

In [4]:
# original
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [5]:
# headers with changes
autos.columns = ['crawled_date', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'reg_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'reg_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created_date', 'nrofpictures', 'postal_code',
       'lastseen_date']
autos.columns.str.lower()

Index(['crawled_date', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'reg_year', 'gearbox', 'power_ps', 'model', 'odometer',
       'reg_month', 'fuel_type', 'brand', 'unrepaired_damage',
       'ad_created_date', 'nrofpictures', 'postal_code', 'lastseen_date'],
      dtype='object')

## 2.2 Exploring data to clean

Initially we will look for:

- Text columns where all or almost all values are the same. These can often be dropped as they don't have useful information for analysis.
- Examples of numeric data stored as text which can be cleaned and converted.

The following methods are helpful for exploring the data:

- ```DataFrame.describe()``` (with include='all' to get both categorical and numeric columns)
- ``` Dataframe.nunique()``` This method returns the count of unique values in the specified axis
- ```Series.value_counts()``` to generate frequency tables
    - Because ```Series.value_counts()``` returns a series, we can use ```Series.sort_index()``` with ascending= True or False to view the highest and lowest values with their counts (can also chain to head() here)
    - To include missing values in the distribution and to use percentages instead of counts, chain the ```Series.value_counts(normalize=True, dropna=False)``` method.
- ```Series.head()``` if any columns need a closer look
- ```Series.unique().shape``` to see how many unique values
- ```Series.describe()``` to view min/max/median/mean etc
- ```DataFrame.isnull().sum()``` count of NaN values for every column

In [6]:
# first let's look at the whole dataset
print(autos.info())
autos.describe(include='all')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   crawled_date       50000 non-null  object
 1   name               50000 non-null  object
 2   seller             50000 non-null  object
 3   offer_type         50000 non-null  object
 4   price              50000 non-null  object
 5   ab_test            50000 non-null  object
 6   vehicle_type       44905 non-null  object
 7   reg_year           50000 non-null  int64 
 8   gearbox            47320 non-null  object
 9   power_ps           50000 non-null  int64 
 10  model              47242 non-null  object
 11  odometer           50000 non-null  object
 12  reg_month          50000 non-null  int64 
 13  fuel_type          45518 non-null  object
 14  brand              50000 non-null  object
 15  unrepaired_damage  40171 non-null  object
 16  ad_created_date    50000 non-null  objec

crawled_date         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-27 22:55:05  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type      reg_year  gearbox      power_ps  model   odometer  \
count         44905  50000.000000    47320  50000.000000  47242      50000   
unique            8           NaN        2           NaN    245         13   
top       limousine           NaN  manuell           NaN   golf  150,000km   
freq          12859           NaN    36993           NaN   4024      32424   
mean            NaN   2005.073280      NaN    116.355920    NaN        NaN   
std             NaN    105.712813      NaN    209.216627    NaN        NaN   
min             NaN   1000.000000      NaN      0.000000    NaN        NaN   
25%             NaN   1999.000000      NaN     70.000000    NaN        NaN   
50%             NaN   2003.000000      NaN    105.000000    NaN        NaN   
75%             NaN   2008.000000      NaN    150.000000    NaN        NaN   
max             NaN   9999.000000      NaN  17700.000000    NaN        NaN   

           reg_month fuel_type       brand unrepaired_damage  \
count   50000.000000     45518       50000             40171   
unique           NaN         7          40                 2   
top              NaN    benzin  volkswagen              nein   
freq             NaN     30107       10687             35232   
mean        5.723360       NaN         NaN               NaN   
std         3.711984       NaN         NaN               NaN   
min         0.000000       NaN         NaN               NaN   
25%         3.000000       NaN         NaN               NaN   
50%         6.000000       NaN         NaN               NaN   
75%         9.000000       NaN         NaN               NaN   
max        12.000000       NaN         NaN               NaN   

            ad_created_date  nrofpictures   postal_code        lastseen_date  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

### 2.2.1 Columns with unique values
Let's take a closer look at columns with only 2 and less unique values. We will generate frequency tables for them and see if we can drop any of them.

In [7]:
autos.nunique()

crawled_date         48213
name                 38754
seller                   2
offer_type               2
price                 2357
ab_test                  2
vehicle_type             8
reg_year                97
gearbox                  2
power_ps               448
model                  245
odometer                13
reg_month               13
fuel_type                7
brand                   40
unrepaired_damage        2
ad_created_date         76
nrofpictures             1
postal_code           7014
lastseen_date        39481
dtype: int64

In [8]:
# lets generate frequency tables for columns that have two and less unique values to have a closer look
for r in autos.columns:
    if autos[r].nunique() <= 2:
        print(autos[r].value_counts(dropna=False))
        print('\n')

privat        49999
gewerblich        1
Name: seller, dtype: int64


Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64


test       25756
control    24244
Name: ab_test, dtype: int64


manuell      36993
automatik    10327
NaN           2680
Name: gearbox, dtype: int64


nein    35232
NaN      9829
ja       4939
Name: unrepaired_damage, dtype: int64


0    50000
Name: nrofpictures, dtype: int64




### 2.2.2 Columns with NaN values
Let's take a closer look at columns with NaN values and see if we can correct some of them.

In [9]:
autos.isnull().sum()

crawled_date            0
name                    0
seller                  0
offer_type              0
price                   0
ab_test                 0
vehicle_type         5095
reg_year                0
gearbox              2680
power_ps                0
model                2758
odometer                0
reg_month               0
fuel_type            4482
brand                   0
unrepaired_damage    9829
ad_created_date         0
nrofpictures            0
postal_code             0
lastseen_date           0
dtype: int64

In [10]:
# generate frequency tables for columns with NaN
for r in autos.columns:
    if autos[r].isnull().sum() > 0:
        print(autos[r].value_counts(dropna=False))
        print('\n')

limousine     12859
kleinwagen    10822
kombi          9127
NaN            5095
bus            4093
cabrio         3061
coupe          2537
suv            1986
andere          420
Name: vehicle_type, dtype: int64


manuell      36993
automatik    10327
NaN           2680
Name: gearbox, dtype: int64


golf          4024
andere        3528
3er           2761
NaN           2758
polo          1757
              ... 
kalina           2
b_max            1
i3               1
200              1
rangerover       1
Name: model, Length: 246, dtype: int64


benzin     30107
diesel     14567
NaN         4482
lpg          691
cng           75
hybrid        37
andere        22
elektro       19
Name: fuel_type, dtype: int64


nein    35232
NaN      9829
ja       4939
Name: unrepaired_damage, dtype: int64




### 2.2.3 Observations for which data can be cleaned
1. **Drop Columns** 
    1. two columns have only 2 unique values and can be dropped as they have only one row each that is adding 2nd unique value
        - seller
        - offer_type
    2. nrofpictures has one unique value so it can be dropped
2. **Convert to numbers** by removing string from value
    1. price
    2. odometer
3. **Convert to dates**
    1. crawled_date
    2. ad_created_date
    3. lastseen_date
4. **Convert NaN values**
    1. gearbox - fill NaN values with most popular value 'Manaul'
    2. unrepaired_damage - fill NaN values with most popular value 'nein'

## 2.3 Cleaning data

We will execute only some data cleaning from total observations.

We will use following functions:
- ```DataFrame.rename({"old_col_name": "new_col_name"}, axis=1, inplace=True)``` to rename the column
- ``` mapping_dict = {'Android': 'Android','Chrome OS': 'Chrome OS'}
DataFrame["column_name"] = DataFrame["column_name"].map(mapping_dict)``` Replacing Values Using A Mapping Dictionary. Note that any value that is not in dictionary will be replace by NaN
-  ```DataFrame.dropna(axis=0)``` Dropping Missing Values
- ```Series.str.replace('from','to')```
- ```Series.astype(float)``` Converting A String Column To Float
- ```Series.str.split().str[0]``` Extracting Values From Strings (in this example first word due to .str[0])
- When removing outliers, we can do ```df[(df["col"] >= x ) & (df["col"] <= y )]```, but it's more readable to use ```df[df["col"].between(x,y)]```
- `Series.index` attribute to access the labels

### 2.3.1 Converting some columns to Numbers

In [11]:
autos['price'] = (autos['price']
                    .str.replace('$','')
                    .str.replace(',','')
                    .astype(int)
                 )
autos['price'].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64

In [12]:
autos['odometer'] = (autos['odometer']
                    .str.replace('km','')
                    .str.replace(',','')
                    .astype(int)
                 )
autos.rename({"odometer": "odometer_km"}, axis=1, inplace=True)
autos["odometer_km"].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64

### 2.3.2 Cleaning outliers (MIN/MAX) Number columns
Analyze the columns with number values using minimum and maximum values and look for any values that look unrealistically high or low (outliers) that we might want to remove.

We know by using ```DataFrames.info()``` that there are 7 colums that are integers. Lets take a look at their min/max and generate frequency tables.

In [13]:
#genrerate min/max
autos.describe()

price      reg_year      power_ps    odometer_km     reg_month  \
count  5.000000e+04  50000.000000  50000.000000   50000.000000  50000.000000   
mean   9.840044e+03   2005.073280    116.355920  125732.700000      5.723360   
std    4.811044e+05    105.712813    209.216627   40042.211706      3.711984   
min    0.000000e+00   1000.000000      0.000000    5000.000000      0.000000   
25%    1.100000e+03   1999.000000     70.000000  125000.000000      3.000000   
50%    2.950000e+03   2003.000000    105.000000  150000.000000      6.000000   
75%    7.200000e+03   2008.000000    150.000000  150000.000000      9.000000   
max    1.000000e+08   9999.000000  17700.000000  150000.000000     12.000000   

       nrofpictures   postal_code  
count       50000.0  50000.000000  
mean            0.0  50813.627300  
std             0.0  25779.747957  
min             0.0   1067.000000  
25%             0.0  30451.000000  
50%             0.0  49577.000000  
75%             0.0  71540.000000  
max             0.0  99998.000000

In [14]:
#generate Unique values for int col
for r in autos.columns:
    if autos[r].dtype == 'int64':
        print(r)
        print(autos[r].unique().shape[0])
            

price
2357
reg_year
97
power_ps
448
odometer_km
13
reg_month
13
nrofpictures
1
postal_code
7014


In [15]:
# generate frequency tables for columns with integers
for r in autos.columns:
    if autos[r].dtype == 'int64':
        print(autos[r].value_counts(dropna=False).sort_index(ascending=False))
        print('\n')

99999999       1
27322222       1
12345678       3
11111111       2
10000000       1
            ... 
5              2
3              1
2              3
1            156
0           1421
Name: price, Length: 2357, dtype: int64


9999    4
9996    1
9000    2
8888    1
6200    1
       ..
1800    2
1500    1
1111    1
1001    1
1000    1
Name: reg_year, Length: 97, dtype: int64


17700       1
16312       1
16011       1
15016       1
15001       1
         ... 
4           4
3           3
2           2
1           5
0        5500
Name: power_ps, Length: 448, dtype: int64


150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
20000       784
10000       264
5000        967
Name: odometer_km, dtype: int64


12    3447
11    3360
10    3651
9     3389
8     3191
7     3949
6     4368
5     4107
4     4102
3     5071
2     3008
1     3282
0     5075
Name: reg_month, dtype: int64


0  

We can **conclude** that:

Incorect values as users could enter number manually:
- price > has 2357 unique values with some really unrealistically high values and 2.8% records with zero
    - **we will remove** Given that eBay is an auction site, there could legitimately be items where the opening bid is 1 We will keep the 1 items, but remove anything above 350,000, since it seems that prices increase steadily to that number and then jump up to less realistic numbers.
- reg_year > has 97 unique values with some unreallistic years on both ends
    - **we will remove** rows where year is less 1900 and more 2020. But we will also make sure that % of data we removing is small.
- power_ps > has 448 unique values with few unreallistically high values and 10% of records zeros
- reg_month > has 13 unique values as there are 5075 records with zero

Good or not clear:
- postal_code > has 7014 unique values but not clear which once are correct
- odometer_km > has 13 uniqe values looks like users had dropdown list for this column

In [16]:
#Removing records due to wrong price
old_shape = autos.shape[0]
autos = autos[autos["price"].between(1,351000)]
removed_records = old_shape - autos.shape[0]
print('We removed {} records as they had wrong price. {} records left Dataset.'
      .format(removed_records,autos.shape[0]))

We removed 1435 records as they had wrong price. 48565 records left Dataset.


In [17]:
#Assessing % records to remove due to bad year
bad_year_rows = (~autos["reg_year"].between(1900,2016)).sum() #do not forget where to put brackets (condition).sum
prct_to_remove = bad_year_rows / autos.shape[0] * 100
print('We will remove {:.2f}% of records as their Registration year is wrong.'
      .format(prct_to_remove))

We will remove 3.88% of records as their Registration year is wrong.


In [18]:
#removing records due to wrong year
old_shape = autos.shape[0]
autos = autos[autos["reg_year"].between(1900,2016)]
removed_records = old_shape - autos.shape[0]
print('We removed {} records due to wrong year. {} records left Dataset.'
      .format(removed_records,autos.shape[0]))

We removed 1884 records due to wrong year. 46681 records left Dataset.


### 2.3.3 Cleaning Data columns

There are 5 columns that should represent date values. Some of these columns were created by the crawler, some came from the website itself. We can differentiate by referring to the data dictionary:

Below two columns represent date in Int64 format:
- `reg_month`: from the website
- `reg_year`: from the website

We will focus only on Date columns:
- `crawled_date`: added by the crawler
- `lastseen_date`: added by the crawler
- `ad_created_date`: from the website

Let's first understand how the values in the three string columns are formatted. These columns all represent full timestamp values, like so:

In [19]:
autos[['crawled_date','ad_created_date','lastseen_date']][0:10]

crawled_date      ad_created_date        lastseen_date
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50
5  2016-03-21 13:47:45  2016-03-21 00:00:00  2016-04-06 09:45:21
6  2016-03-20 17:55:21  2016-03-20 00:00:00  2016-03-23 02:48:59
7  2016-03-16 18:55:19  2016-03-16 00:00:00  2016-04-07 03:17:32
8  2016-03-22 16:51:34  2016-03-22 00:00:00  2016-03-26 18:18:10
9  2016-03-16 13:47:02  2016-03-16 00:00:00  2016-04-06 10:46:35

We will not do analysis of time in this project so we will just drop time information for above 3 columns.

In [20]:
autos['crawled_date'] = autos['crawled_date'].str[:10]
autos['ad_created_date'] = autos['ad_created_date'].str[:10]
autos['lastseen_date'] = autos['lastseen_date'].str[:10]

In [21]:
autos[['crawled_date','ad_created_date','lastseen_date']][0:10]

crawled_date ad_created_date lastseen_date
0   2016-03-26      2016-03-26    2016-04-06
1   2016-04-04      2016-04-04    2016-04-06
2   2016-03-26      2016-03-26    2016-04-06
3   2016-03-12      2016-03-12    2016-03-15
4   2016-04-01      2016-04-01    2016-04-01
5   2016-03-21      2016-03-21    2016-04-06
6   2016-03-20      2016-03-20    2016-03-23
7   2016-03-16      2016-03-16    2016-04-07
8   2016-03-22      2016-03-22    2016-03-26
9   2016-03-16      2016-03-16    2016-04-06

# Part 3 - Analyze Data

## 3.1 Explore Date periods of dataset

### 3.1.1 Crawler dates
Lets check for which period crawler collected data.


In [22]:
(autos["crawled_date"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

2016-03-05    0.025192
2016-03-06    0.014160
2016-03-07    0.036246
2016-03-08    0.033547
2016-03-09    0.033247
2016-03-10    0.032240
2016-03-11    0.032454
2016-03-12    0.036824
2016-03-13    0.015874
2016-03-14    0.036332
2016-03-15    0.034361
2016-03-16    0.029498
2016-03-17    0.031790
2016-03-18    0.012810
2016-03-19    0.034661
2016-03-20    0.038024
2016-03-21    0.037317
2016-03-22    0.032840
2016-03-23    0.032197
2016-03-24    0.029477
2016-03-25    0.031512
2016-03-26    0.032069
2016-03-27    0.030783
2016-03-28    0.034597
2016-03-29    0.034104
2016-03-30    0.033804
2016-03-31    0.031790
2016-04-01    0.033804
2016-04-02    0.035561
2016-04-03    0.038774
2016-04-04    0.036610
2016-04-05    0.013003
2016-04-06    0.003085
2016-04-07    0.001414
Name: crawled_date, dtype: float64

Looks like the site was crawled daily over roughly a one month period in March and April 2016. The distribution of listings crawled on each day is roughly uniform.

### 3.1.2 Last Seen dates

In [23]:
(autos["lastseen_date"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

2016-03-05    0.001071
2016-03-06    0.004113
2016-03-07    0.005377
2016-03-08    0.007476
2016-03-09    0.009768
2016-03-10    0.010690
2016-03-11    0.012382
2016-03-12    0.023757
2016-03-13    0.008654
2016-03-14    0.012660
2016-03-15    0.016002
2016-03-16    0.016281
2016-03-17    0.028084
2016-03-18    0.007219
2016-03-19    0.015617
2016-03-20    0.020629
2016-03-21    0.020587
2016-03-22    0.020844
2016-03-23    0.018359
2016-03-24    0.019687
2016-03-25    0.018937
2016-03-26    0.016795
2016-03-27    0.015638
2016-03-28    0.020694
2016-03-29    0.022086
2016-03-30    0.024614
2016-03-31    0.023628
2016-04-01    0.022943
2016-04-02    0.024657
2016-04-03    0.025149
2016-04-04    0.024121
2016-04-05    0.125404
2016-04-06    0.223324
2016-04-07    0.132752
Name: lastseen_date, dtype: float64

The crawler recorded the date it last saw any listing, which allows us to determine on what day a listing was removed, presumably because the car was sold.
The last three days contain a disproportionate amount of 'last seen' values. Given that these are 6-10x the values from the previous days, it's unlikely that there was a massive spike in sales, and more likely that these values are to do with the crawling period ending and don't indicate car sales.

### 3.1.3 Ad dates

In [25]:
print(autos["ad_created_date"].str[:10].unique().shape)
(autos["ad_created_date"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

(74,)


2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.039009
2016-04-04    0.036953
2016-04-05    0.011782
2016-04-06    0.003170
2016-04-07    0.001264
Name: ad_created_date, Length: 74, dtype: float64

There is a large variety of ad created dates. Most fall within 1-2 months of the listing date, but a few are quite old, with the oldest at around 9 months.

## 3.2 Exploring Price by Brand

In [26]:
autos["brand"].value_counts(normalize=True)

volkswagen        0.211264
bmw               0.110045
opel              0.107581
mercedes_benz     0.096463
audi              0.086566
ford              0.069900
renault           0.047150
peugeot           0.029841
fiat              0.025642
seat              0.018273
skoda             0.016409
nissan            0.015274
mazda             0.015188
smart             0.014160
citroen           0.014010
toyota            0.012703
hyundai           0.010025
sonstige_autos    0.009811
volvo             0.009147
mini              0.008762
mitsubishi        0.008226
honda             0.007840
kia               0.007069
alfa_romeo        0.006641
porsche           0.006127
suzuki            0.005934
chevrolet         0.005698
chrysler          0.003513
dacia             0.002635
daihatsu          0.002506
jeep              0.002271
subaru            0.002142
land_rover        0.002099
saab              0.001649
jaguar            0.001564
daewoo            0.001500
trabant           0.001392
r

German manufacturers represent four out of the top five brands, almost 50% of the overall listings. Volkswagen is by far the most popular brand, with approximately double the cars for sale of the next two brands combined.
There are lots of brands that don't have a significant percentage of listings, so we will limit our analysis to brands representing more than 5% of total listings.

In [27]:
#finding most common brands
brand_counts = autos["brand"].value_counts(normalize=True)
common_brands = brand_counts[brand_counts > .05].index
print(common_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [28]:
#finding average prices for top brands
brand_mean_prices = {}

for brand in common_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_price = brand_only["price"].mean()
    brand_mean_prices[brand] = int(mean_price)

brand_mean_prices

{'volkswagen': 5402,
 'bmw': 8332,
 'opel': 2975,
 'mercedes_benz': 8628,
 'audi': 9336,
 'ford': 3749}

Of the top 5 brands, there is a distinct price gap:
- Audi, BMW and Mercedes Benz are more expensive
- Ford and Opel are less expensive
- Volkswagen is in between - this may explain its popularity, it may be a 'best of 'both worlds' option.

## 3.3 Exploring Mileage

For the top 6 brands, let's use aggregation to understand the average mileage for those cars and if there's any visible link with mean price. While our natural instinct may be to display both aggregated series objects and visually compare them, this has a few limitations:

- it's difficult to compare more than two aggregate series objects if we want to extend to more columns
- we can't compare more than a few rows from each series object
- we can only sort by the index (brand name) of both series objects so we can easily make visual comparisons

Instead, we can combine the data from both series objects into a single dataframe (with a shared index) and display the dataframe directly. To do this, we'll need to learn two pandas methods:

- [pandas series constructor](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.html)
- [pandas dataframe constructor](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.html)

Here's an example of the series constructor that uses the brand_mean_prices dictionary.The keys in the dictionary became the index in the series object. We can then create a single-column dataframe from this series object. We need to use the columns parameter when calling the dataframe constructor (which accepts an array-like object) to specify the column name (or the column name will be set to 0 by default).

In [32]:
bmp_series = pd.Series(brand_mean_prices)
pd.DataFrame(bmp_series, columns=["mean_price"])

mean_price
volkswagen           5402
bmw                  8332
opel                 2975
mercedes_benz        8628
audi                 9336
ford                 3749

In [33]:
brand_mean_mileage = {}

for brand in common_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_mileage = brand_only["odometer_km"].mean()
    brand_mean_mileage[brand] = int(mean_mileage)

mean_mileage = pd.Series(brand_mean_mileage).sort_values(ascending=False)
mean_prices = pd.Series(brand_mean_prices).sort_values(ascending=False)

In [36]:
brand_info = pd.DataFrame(mean_mileage,columns=['mean_mileage'])
brand_info["mean_price"] = mean_prices
brand_info

mean_mileage  mean_price
bmw                  132572        8332
mercedes_benz        130788        8628
opel                 129310        2975
audi                 129157        9336
volkswagen           128707        5402
ford                 124266        3749

The range of car mileages does not vary as much as the prices do by brand, instead all falling within 10% for the top brands. There is a slight trend to the more expensive vehicles having higher mileage, with the less expensive vehicles having lower mileage.

# NEXT STEPS

Data cleaning next steps:
- Identify categorical data that uses german words, translate them and map the values to their english counterparts
- Convert the dates to be uniform numeric data, so "2016-03-21" becomes the integer 20160321.
- See if there are particular keywords in the name column that you can extract as new columns

Analysis next steps:
- Find the most common brand/model combinations
- Split the odometer_km into groups, and use aggregation to see if average prices follows any patterns based on the mileage.
- How much cheaper are cars with damage than their non-damaged counterparts?